File: area_scan_3d_camera/Util/SetScanningParameters/SetScanningParameters.cpp
/*******************************************************************************
 *BSD 3-Clause License
 *
 *Copyright (c) 2016-2024, Mech-Mind Robotics
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are met:
 *
 *1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *3. Neither the name of the copyright holder nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 *FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 *OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

/*
With this sample, you can set the parameters in the "3D Parameters", "2D Parameters", and "ROI"
categories.
*/
#include "area_scan_3d_camera/Camera.h"
#include "area_scan_3d_camera/api_util.h"
#include "area_scan_3d_camera/parameters/Scanning3D.h"
#include "area_scan_3d_camera/parameters/Scanning2D.h"

int main()
{
    // List all available cameras and connect to a camera by the displayed index.
    mmind::eye::Camera camera;
    if (!findAndConnect(camera))
        return -1;

    // Obtain the basic information of the connected camera.
    mmind::eye::CameraInfo cameraInfo;
    showError(camera.getCameraInfo(cameraInfo));
    printCameraInfo(cameraInfo);

    mmind::eye::UserSet& currentUserSet = camera.currentUserSet();

    // Set the exposure times for acquiring depth information.
    showError(currentUserSet.setFloatArrayValue(
        mmind::eye::scanning3d_setting::ExposureSequence::name, std::vector<double>{5, 10}));

    // Obtain the current exposure times for acquiring depth information to check if the setting was
    // successful.
    std::vector<double> exposureSequence;
    showError(currentUserSet.getFloatArrayValue(
        mmind::eye::scanning3d_setting::ExposureSequence::name, exposureSequence));
    std::cout << "3D scanning exposure multiplier : " << exposureSequence.size() << "."
              << std::endl;
    for (size_t i = 0; i < exposureSequence.size(); i++) {
        std::cout << "3D scanning exposure time " << i + 1 << ": " << exposureSequence[i] << " ms."
                  << std::endl;
    }

    // Set the ROI for the depth map and point cloud, and then obtain the parameter value to check
    // if the setting was successful.
    mmind::eye::ROI roi(0, 0, 500, 500);
    showError(currentUserSet.setRoiValue(mmind::eye::scanning3d_setting::ROI::name, roi));
    mmind::eye::ROI curRoi;
    showError(currentUserSet.getRoiValue(mmind::eye::scanning3d_setting::ROI::name, curRoi));
    std::cout << "3D scanning ROI topLeftX: " << curRoi.upperLeftX
              << ", topLeftY: " << curRoi.upperLeftY << ", width: " << curRoi.width
              << ", height: " << curRoi.height << std::endl;

    // Set the exposure mode and exposure time for capturing the 2D image, and then obtain the
    // parameter values to check if the setting was successful.
    showError(currentUserSet.setEnumValue(
        mmind::eye::scanning2d_setting::ExposureMode::name,
        static_cast<int>(mmind::eye::scanning2d_setting::ExposureMode::Value::Timed)));

    showError(
        currentUserSet.setFloatValue(mmind::eye::scanning2d_setting::ExposureTime::name, 100));

    int exposureMode2D = 0;
    showError(currentUserSet.getEnumValue(mmind::eye::scanning2d_setting::ExposureMode::name,
                                          exposureMode2D));
    double scan2DExposureTime = 0;
    showError(currentUserSet.getFloatValue(mmind::eye::scanning2d_setting::ExposureTime::name,
                                           scan2DExposureTime));
    std::cout << "2D scanning exposure mode: " << exposureMode2D
              << ", exposure time: " << scan2DExposureTime << " ms." << std::endl;

    // Save all the parameter settings to the currently selected user sets.
    showError(currentUserSet.saveAllParametersToDevice());
    std::cout << "Save all parameters to the current user set." << std::endl;

    camera.disconnect();
    std::cout << "Disconnected from the camera successfully." << std::endl;
    return 0;
}

File: area_scan_3d_camera/util/set_scanning_parameters.py
# With this sample, you can set the parameters in the "3D Parameters", "2D Parameters", and "ROI" categories.

from mecheye.shared import *
from mecheye.area_scan_3d_camera import *
from mecheye.area_scan_3d_camera_utils import find_and_connect, print_camera_info


class SetScanningParameters(object):
    def __init__(self):
        self.camera = Camera()

    def set_scanning_parameters(self):

        # Obtain the basic information of the connected camera.
        cameraInfo = CameraInfo()
        show_error(self.camera.get_camera_info(cameraInfo))
        print_camera_info(cameraInfo)

        # Obtain the name of the currently selected user set.
        current_user_set = self.camera.current_user_set()
        error, user_set_name = current_user_set.get_name()
        show_error(error)
        print("\ncurrent_user_set: " + user_set_name)

        # Set the exposure times for acquiring depth information.
        error = current_user_set.set_float_array_value(
            Scanning3DExposureSequence.name, [5])
        # error = current_user_set.set_float_array_value(
        #     Scanning3DExposureSequence.name, [5, 10])
        show_error(error)
        error, exposure_sequence = current_user_set.get_float_array_value(
            Scanning3DExposureSequence.name)
        show_error(error)
        print("\nThe 3D scanning exposure multiplier: {}".format(
            len(exposure_sequence)))
        for i in exposure_sequence:
            print("3D scanning exposure time: {}".format(i))

        # Set the ROI for the depth map and point cloud, and then obtain the parameter values for checking.
        roi = ROI(0, 0, 500, 500)
        error = current_user_set.set_roi_value(Scanning3DROI.name, roi)
        show_error(error)
        error, roi = current_user_set.get_roi_value(Scanning3DROI.name)
        show_error(error)
        print("\n3D scanning ROI topLeftX: {}, topLeftY: {}, width: {}, height: {}".
              format(roi.upper_left_x, roi.upper_left_y, roi.width, roi.height))

        # Set the exposure mode and exposure time for capturing the 2D image, and then obtain the
        # parameter values to check if the setting was successful.
        exposure_mode_2d = Scanning2DExposureMode.Value_Timed
        error = current_user_set.set_enum_value(
            Scanning2DExposureMode.name, exposure_mode_2d)
        show_error(error)
        exposure_time_2d = 100
        error = current_user_set.set_float_value(
            Scanning2DExposureTime.name, exposure_time_2d)
        show_error(error)

        # The DEEP and LSR series also provide a "Scan2DPatternRoleExposureMode" parameter for
        # adjusting the exposure mode for acquiring the 2D images (depth source). Uncomment the
        # following lines to set this parameter to "Timed".
        depth_source_exposure_mode_2d = Scanning2DDepthSourceExposureMode.Value_Timed
        error = current_user_set.set_enum_value(
            Scanning2DDepthSourceExposureMode.name, depth_source_exposure_mode_2d)
        show_error(error)

        # You can also use the projector for supplemental light when acquiring the 2D image / 2D images
        # (depth source).
        # Models other than the DEEP and LSR series: Uncomment the following lines to set the exposure
        # mode to "Flash" for supplemental light.
        # exposure_mode_2d = Scanning2DExposureMode.Value_Flash
        # error = current_user_set.set_enum_value(
        #     Scanning2DExposureMode.name, exposure_mode_2d)
        # show_error(error)

        # DEEP and LSR series: Uncomment the following lines to set the exposure mode to "Flash" for
        # supplemental light.
        # depth_source_exposure_mode_2d = Scanning2DDepthSourceExposureMode.Value_Flash
        # error = current_user_set.set_enum_value(
        #     Scanning2DDepthSourceExposureMode.name, depth_source_exposure_mode_2d)
        # show_error(error)

        # The following models also provide a "FlashAcquisitionMode" when using the flash exposure
        # mode: DEEP, DEEP-GL, LSR S/L/XL, LSR S-GL/L-GL/XL-GL, PRO XS/S/M, PRX XS-GL/S-GL/M-GL, NANO, NANO-GL, NANO ULTRA, NANO ULTRA-GL. Uncomment the following lines to set
        # the "FlashAcquisitionMode" parameter to "Responsive".
        # flash_acquisition_mode_2d=Scanning2DFlashAcquisitionMode.Value_Responsive
        # error = current_user_set.set_enum_value(
        #     Scanning2DFlashAcquisitionMode.name, flash_acquisition_mode_2d)
        # show_error(error)

        # When using the responsive acquisition mode, you can adjust the exposure time for the flash
        # exposure mode. Uncomment the following lines to set the exposure time to 20 ms.
        # flash_exposure_time_2d = 20
        # error = current_user_set.set_float_value(
        #     Scanning2DFlashExposureTime.name, flash_exposure_time_2d)
        # show_error(error)

        # Uncomment the following lines to check the values of the "FlashAcquisitionMode" and "FlashExposureTime" parameters.
        # error, flash_acquisition_mode_2d = current_user_set.get_enum_value_string(
        #     Scanning2DFlashAcquisitionMode.name)
        # show_error(error)
        # error, flash_exposure_time_2d = current_user_set.get_float_value(
        #     Scanning2DFlashExposureTime.name)
        # show_error(error)
        # print("\n2D scanning flash acquisition mode: {}, flash exposure time: {}".
        #       format(flash_acquisition_mode_2d, flash_exposure_time_2d))

        # Uncomment the following line to set the camera gain when capturing 2D images.
        # For DEEP and LSR series cameras, the camera gain for capturing 2D images (texture) can be
        # adjusted by setting the "Scan2DGain" parameter.
        # For other camera series, the camera gain for capturing 2D images can be adjusted by
        # setting the "Scan2DGain" parameter when the exposure mode is set to fixed exposure, auto
        # exposure, HDR, or flash mode, and the flash acquisition mode is set to
        # responsive.
        # print("\n2D image gain: ", scan2DGain)value(Scanning2DGain.name,2.0))
        # show_error(error)et_float_value(Scanning2DGain.name)
        # error, scan2DGain = current_user_set.get_float_value(Scanning2DGain.name)
        # show_error(current_user_set.set_float_value(Scanning2DGain.name,2.6))

        error, exposure_mode_2d = current_user_set.get_enum_value_string(
            Scanning2DExposureMode.name)
        show_error(error)
        error, exposure_time_2d = current_user_set.get_float_value(
            Scanning2DExposureTime.name)
        show_error(error)
        print("\n2D scanning exposure mode enum: {}, exposure time: {}".
              format(exposure_mode_2d, exposure_time_2d))

        # Save all the parameter settings to the currently selected user set.
        success_message = "\nSave the current parameter settings to the selected user set."
        show_error(current_user_set.save_all_parameters_to_device(), success_message)

    def main(self):
        if find_and_connect(self.camera):
            self.set_scanning_parameters()
            self.camera.disconnect()
            print("Disconnected from the camera successfully.")


if __name__ == '__main__':
    a = SetScanningParameters()
    a.main()
